Mathlib/Tactic/Simproc/FinsetInterval.lean

```diff
@@ -0,0 +1,188 @@
+    Icc m n = insert m s := by rw [← hs, Nat.Icc_insert_succ_left (by simpa using hmn)]
+
+private lemma Ico_eq_of_Icc_pred_eq (hn : n ≠ 0) (hs : Icc m (n - 1) = s) : Ico m n = s := by
+  rw [← hs, Icc_sub_one_right_eq_Ico_of_not_isMin (by simpa)]
+
+private lemma Ico_zero (m : ℕ) : Ico m 0 = ∅ := by simp
+
+private lemma Ioc_eq_of_Icc_succ_eq (hs : Icc (m + 1) n = s) : Ioc m n = s := by
+  rw [← hs, Icc_add_one_left_eq_Ioc]
+
+private lemma Ioo_eq_of_Icc_succ_pred_eq (hs : Icc (m + 1) (n - 1) = s) : Ioo m n = s := by
+  rw [← hs, ← Icc_add_one_sub_one_eq_Ioo]
+
+private lemma Iic_eq_of_Icc_zero_eq (hs : Icc 0 n = s) : Iic n = s := hs
+
+private lemma Iio_eq_of_Icc_zero_pred_eq (hn : n ≠ 0) (hs : Icc 0 (n - 1) = s) : Iio n = s :=
+  Ico_eq_of_Icc_pred_eq hn hs
+
+private lemma Iio_zero : Iio 0 = ∅ := by simp
+
+end lemmas
+
+/-- Given natural numbers `m` and `n`, returns `(s, ⊢ Finset.Icc m n = s)`. -/
+def evalFinsetIccNat (m n : ℕ) (em en : Q(ℕ)) :
+    MetaM ((s : Q(Finset ℕ)) × Q(.Icc $em $en = $s)) := do
+  -- If `m = n`, then `Icc m n = {m}`. We handle this case separately because `insert m ∅` is
+  -- not syntactically `{m}`.
+  if m = n then
+    have : $em =Q $en := ⟨⟩
+    return ⟨q({$em}), q(Icc_self _)⟩
+  -- If `m < n`, then `Icc m n = insert m (Icc m n)`.
+  else if m < n then
+    let hmn : Q(Nat.ble $em $en = true) := (q(Eq.refl true) :)
+    have em' : Q(ℕ) := mkNatLitQ (m + 1)
+    have : $em' =Q $em + 1 := ⟨⟩
+    let ⟨s, hs⟩ ← evalFinsetIccNat (m + 1) n em' en
+    return ⟨q(insert $em $s), q(Icc_eq_insert_of_Icc_succ_eq $hmn $hs)⟩
+  -- Else `n < m` and `Icc m n = ∅`.
+  else
+    let hnm : Q(Nat.blt $en $em = true) := (q(Eq.refl true) :)
+    return ⟨q(∅), q(Icc_eq_empty_of_lt $hnm)⟩
+
+end Mathlib.Tactic.Simp
+
+open Mathlib.Tactic.Simp
+
+/-!
+Note that these simprocs are not made simp to avoid simp blowing up on goals containing things of
+the form `Iic (2 ^ 1024)`.
+-/
+namespace Finset
+
+/-- Simproc to compute `Finset.Icc a b` when `a b : ℕ`.
+
+**Warnings**:
+* With the standard depth recursion limit, this simproc can compute intervals of size 250 at most.
+* Make sure to exclude `Finset.insert_eq_of_mem` from your simp call when using this simproc. This
+  avoids a quadratic time performance hit. -/
+simproc_decl Icc_nat (Icc _ _) := fun e ↦ do
+  let ⟨1, ~q(Finset ℕ), ~q(Icc (OfNat.ofNat $em) (OfNat.ofNat $en))⟩ ← inferTypeQ e
+    | return .continue
+  let some m := em.rawNatLit? | return .continue
+  let some n := en.rawNatLit? | return .continue
+  let ⟨es, p⟩ ← evalFinsetIccNat m n em en
+  return .done { expr := es, proof? := p }
+
+/-- Simproc to compute `Finset.Ico a b` when `a b : ℕ`.
+
+**Warnings**:
+* With the standard depth recursion limit, this simproc can compute intervals of size 250 at most.
+* Make sure to exclude `Finset.insert_eq_of_mem` from your simp call when using this simproc. This
+  avoids a quadratic time performance hit. -/
+simproc_decl Ico_nat (Ico _ _) := fun e ↦ do
```

Comment:
Could you try to use `ofQ` here?

Reply:
I would love to but it hasn't landed, right?

Reply:
Ah, it `Simproc.ofQ`

Reply:
Sadly, this makes my simprocs not follow the line of sight rule anymore 😦 Any way I can make sure the main branch of my simproc stays unindented?

Reply:
Unfortunately not, but this is not really any different to the typical `| _ => throwUnsupportedSyntax` that appears elsewhere.